### ISSUE TITLE
Microphone not working in Builder in Windows

### ISSUE THREAD

**Issue by peteruhrig** [psychopy/psychopy#2323]
Dear all,

I cannot record audio from the microphone using the relevant features of the Builder.

To reproduce:
1. Create a new experiment in Builder
2. Add text stimulus
3. Add microphone response (e.g. recording from 0 to 2 seconds)
4. Run the experiment

Error message is:
AttributeError: pyo server not created.

The full error message is below.
![image](https://user-images.githubusercontent.com/19432934/53801690-20aa4f80-3f40-11e9-82a9-dcb4c2ad9ab6.png)

Tested in Windows 7 and Windows 10 with PsychoPy 3.0.2 and 3.0.5 (32bit because of the QT problem)

Best,
Peter

**Comment by dvbridges** [comment]
Hi @peteruhrig , I am not able to recreate this error using standalone PsychoPy 3.0.4 running on Windows 10. I am using a Maplin YU-34 microphone and my audio hardware settings are as follows:]

```
audio library: ['pyo', 'sounddevice', 'pygame', ]
audio driver: ['Primary Sound', 'ASIO', 'Audigy', ]
audio device: default
```

**Comment by peteruhrig** [comment]
Hi dvbridges,

thanks for the swift reply! These are my settings, too.
Just to be sure:  Do I need to install anything apart from PsychoPy standalone? (We even tried to install portaudio manually, but this does not change the version PsychoPy uses.)

Best,
Peter

**Comment by dvbridges** [comment]
Ok, well perhaps there is an issue with your sound card and drivers? Would be good to know what you are using, and hopefully other users will also share what hardware they are using so we can figure out what is happening.

**Comment by peircej** [comment]
@peteruhrig If you go to the psychopy shell panel (or create a script) and just do 
```python
import psychopy.sound.backend_pyo as pyo
print(pyo.get_intput_devices())
pyo.init()
```
what happens?

**Comment by peteruhrig** [comment]
@peircej 
Thank you. It depends on the machine and PsychoPy version. I'm trying to help my colleague @minnie2112 with this, who in the meantime has managed to get a step further by installing version 3.04 on her Windows 10 and now gets a new error (or rather, a warning) \[also see her comment in [issue 2284](https://github.com/psychopy/psychopy/issues/2284))\].
On my machine (Windows 7, PsychoPy 3.05, 32bit version) I get the following output:

```
Welcome to PsychoPy3!
v3.0.5

######################### Running: C:\temp\mictry_3.py #########################
pyo version 0.9.0 (uses single precision)
(['Microsoft Soundmapper - Input', 'Mikrofon (3- C-Media USB Headph', 'Primärer Soundaufnahmetreiber', 'Mikrofon (3- C-Media USB Headphone Set  )', 'Mikrofon (3- C-Media USB Headphone Set  )', 'Eingang (Realtek HD Audio Line input)', 'Mikrofon (Realtek HD Audio Mic input)', 'Stereomix (Realtek HD Audio Stereo input)', 'Mikrofon (C-Media USB Headphone Set)'], [0, 1, 5, 6, 12, 14, 15, 16, 18])
```
So this looks healthy to me.

Best,
Peter

P.S.: I had to correct "intput" to "input" in your code snippet.

**Comment by minnie2112** [comment]
I am using Psychopy 3.0.4. (Standalone), on Windows 10 (64bit)
The code of @peircej  gave me the following output:
```
##### Running: E:\FAU\PhD dissertation\phd data\naming task\namingtask_lastrun.py #####
pyo version 0.9.0 (uses single precision)
(['Microsoft Sound Mapper - Input', 'Microphone (Conexant SmartAudio', 'Primary Sound Capture Driver', 'Microphone (Conexant SmartAudio HD)', 'Microphone (Conexant SmartAudio HD)', 'Stereo Mix (Conexant HD Stereo Mix)', 'Microphone (Conexant HD Audio capture)'], [0, 1, 4, 5, 9, 10, 12])
1.2063     WARNING     sounddevice audio lib was requested but not loaded: sounddevice not working
```

Thank you for your help in advance.

**Comment by dvbridges** [comment]
For comparison, I am getting:

```
################ Running: C:\Users\lpzdb\Documents\untitled.py #################
pyo version 0.9.0 (uses single precision)
(['Microsoft Sound Mapper - Input', 'Microphone (Realtek High Defini', 'Primary Sound Capture Driver', 'Microphone (Realtek High Definition Audio)', 'Microphone (Realtek High Definition Audio)', 'Line In (Realtek HD Audio Line input)', 'Stereo Mix (Realtek HD Audio Stereo input)', 'Microphone (Realtek HD Audio Mic input)'], [0, 1, 5, 6, 12, 13, 15, 16])
0.4252     WARNING     sounddevice audio lib was requested but not loaded: sounddevice not working
```

**Comment by peteruhrig** [comment]
OK, we are making progress.
@minnie2112 on version 3.0.4 on Windows 10 (64 bit) can run this script and it works. She still gets the warning "sounddevice audio lib was requested but not loaded", but it records nonetheless.
On my version 3.0.5 (32 bit) on Windows 7 I get the pyo error:
![image](https://user-images.githubusercontent.com/19432934/53971109-7f1c2d00-40fc-11e9-858d-fd68bf4db799.png)
Best wishes,
Peter

**Comment by peteruhrig** [comment]
I just downloaded and installed 3.0.6. The error message for recording from the builder has changed now:
![image](https://user-images.githubusercontent.com/19432934/54031485-28236000-41af-11e9-8031-0532466c98b1.png)

Running Demo -> Hardware -> testSoundLatency.py results in the following error message:
```
##### Running: C:\Program Files (x86)\PsychoPy3\lib\site-packages\psychopy\demos\coder\hardware\testSoundLatency.py #####
pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
Traceback (most recent call last):
  File "C:\Program Files (x86)\PsychoPy3\lib\site-packages\psychopy\demos\coder\hardware\testSoundLatency.py", line 16, in <module>
    from labjack import u3
ModuleNotFoundError: No module named 'labjack'
```

Here is a minimal example that fails in 3.0.6:

```
from psychopy import voicekey, sound

pass
````
Error message:
```
######################### Running: C:\temp\mictry_4.py #########################
Traceback (most recent call last):
  File "C:\Program Files (x86)\PsychoPy3\lib\site-packages\psychopy\voicekey\__init__.py", line 26, in <module>
    import pyo64 as pyo
ModuleNotFoundError: No module named 'pyo64'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\temp\mictry_4.py", line 1, in <module>
    from psychopy import voicekey, sound
  File "C:\Program Files (x86)\PsychoPy3\lib\site-packages\psychopy\voicekey\__init__.py", line 29, in <module>
    import pyo
ModuleNotFoundError: No module named 'pyo'
```

This is all on Windows 7, 64 bit.

Best,
Peter

**Comment by peircej** [comment]
Ah, this one is trickier. The pyo library is currently only provided as a 32bit lib (I'll try and get them to build the 64bit one and/or try to compile it myself) but for now that means the mic is 32bit installs only on windows.
I'm really hoping to have Psychtoolbox sounds working from within PsychoPy soon (conversely that will be *only* 64 bit!) and that should have superior timing too

**Comment by belangeo** [comment]
@peteruhrig, an encoding mismatch (very usual on Windows) ?

> 'Primärer Soundaufnahmetreiber'

**Comment by ebuyakin** [comment]
Dear All,

I have the same problem with sound recording on Mac with PsychoPy 3.05
The experiment includes just one simple routine with a tex component, microphone component and a keyboard response. Running the experiment gives 'pyo server not created' error in the microphone module.

I have a working internal microphone and I tested pyo module from outside of PsychoPy (using PyCharm) and it works fine. Any advice what shall I try to solve the problem?

Are there any alternatives to using a standard microphone component for recording audio? I tried using pyaudio stream, but it keeps crashing PsychoPy.

Best regards,
Eugene

<img width="736" alt="Screenshot 2019-03-11 at 11 58 49" src="https://user-images.githubusercontent.com/9034484/54123031-8e4df400-43f6-11e9-9a83-d786c0628525.png">

**Comment by schubertt** [comment]
> Ah, this one is trickier. The pyo library is currently only provided as a 32bit lib (I'll try and get them to build the 64bit one and/or try to compile it myself) but for now that means the mic is 32bit installs only on windows.
> I'm really hoping to have Psychtoolbox sounds working from within PsychoPy soon (conversely that will be _only_ 64 bit!) and that should have superior timing too

Any updates on this? I'm also hoping to use pyo in Windows 64bit.

**Comment by peircej** [comment]
I believe there is a beta version of pyo now working on Python3 64bit. I didn't get around to inserting it into the 64bit windows release of PsychoPy 3.1.0 but I'll add it for 3.1.1

**Comment by belangeo** [comment]
Just finished the manylinux1 builds today! It's all on test.pypi:

[https://test.pypi.org/project/pyo/0.9.3/#files](https://test.pypi.org/project/pyo/0.9.3/#files)

**Comment by mirh** [comment]
Sound output works with that (and yes, I forced pyo as the first driver preference)
microphone.AdvAudioCapture() still returns the same error above.
(windows 64 bit)

**Comment by belangeo** [comment]
If the server don't boot, it's probably a system's audio configuration problem. In the sound panel, be sure that there is an available input in the capture tab. Be also sure that it uses the same sampling rate as the output device. If you can set pyo's audio host from psychopy, I recommend directsound host (usually named "Primary Sound Driver") as it seems to be the most reliable for almost every configuration...

**Comment by mirh** [comment]
Wait a minute.. After some tinkering I found out the issues. 
Any audio library preference order that hasn't **pyo** as the \*first\* one, fails with the error above (which is pretty darn funny considering you would expect the pyo error to belong to it, and not to the others). 
And last time I checked (contrarily to David's) the default was https://github.com/psychopy/psychopy/blob/8883b2d080755d3dfe2bfc14706e787fd3a8268c/psychopy/preferences/Windows.spec#L111

Secondarily, I don't really know what "Primary Sound" (nor "Audigy".. we aren't using OpenAL, are we?) would be supposed to do. They seem to be just null. 
(while I understand what ASIO is, but for some reason it doesn't work either, initializing my devices.. but still not recognizing them)
If I add 'directsound' to the driver list, I can finally get microphone to work (on python 64 bit, yes)

**Comment by peircej** [comment]
Thanks for working that one out @mirh. It's always a puzzle when the problem doesn't replicate on  my own machine (which had pyo as default). I can confirm the issue with the sound library. PsychoPy was calling its own `sound.init()` to initialise the sound stream when switching on the mic but 'sound.init()' depends on your sound lib in preferences. This is fixed in #2447

I couldn't get it to fail by setting the "wrong" driver. PsychoPy appears to be correctly selecting a default driver from pyo (e.g. Primary Sound Driver) if the requested ones aren't there, so I couldn't get any setting of that to fail myself.
